scripts/make-leader.sh

```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+source "$(dirname $0)/config.sh"
+
+curl --insecure \
```

Comment:
What's the advantage to handling this as an endpoint here? To me it seems like would be better to configure this via a parameter on startup. Right now theoretically anyone on the network can turn a worker into a leader by calling this which can lead to undefined behavior (e.g. multiple leaders/take over workers by calling heartbeat endpoint on workers)

Reply:
We designate a leader at run time because we cannot do it at compile time: If we start the leader with a different flag then workers, they end up with different image IDs, which prevents key synchronization.

That said, I agree that this is not ideal. Perhaps we can use AWS's Instance Metadata Service for leader designation, i.e., an enclave knows that it's the leader if its domain name contains the string "leader". I'll look into this. (cc @rillian)

Reply:
Ahh that makes sense now! I hadn't considered the impact of image IDs changing. If you want to go really simple with it I'd be fine with establishing a list of public keys that get built into the image/source code and then just using an HTTP signature to authenticate to the API when running the script.

That keeps it simple but also makes it so it's less likely for a random malicious server within the network to call it.

Reply:
@NullHypothesis *Kubernetes in Action* talks about a "Downward API" which can pass metadata to the container, either in env variables or files on a mounted volume. That could be a mechanism to pass configuration data to nitriding, alongside the AWS instance metadata endpoint.

The startup shim running on the ec2 node, but not inside the enclave, could use either mechanism to determine configuration and then make the call to activate the leader in necessary. That moves the config out of the shim image as well, which is nice.

If we prefer nitriding pull config on startup, the shim could run a local-only server to answer the request from the enclave. That helps with the public endpoint issue.

Whether push or pull, we could still sign the configuration payload.

Reply:
All that aside, if nitriding could determine it's calling itself (nonce? uuid?) when trying to contact the leader, it could use that to assume leadership based on external network configuration and the dns hostname for the leader we're already including in the enclave image.